src/State/Hub.php

```diff
@@ -0,0 +1,228 @@
+     * since the last call to {@see Hub::pushScope} are discarded.
+     *
+     * @return bool
+     */
+    public function popScope(): bool
+    {
+        if (1 === \count($this->stack)) {
+            return false;
+        }
+
+        return null !== \array_pop($this->stack);
+    }
+
+    /**
+     * Creates a new scope with and executes the given operation within. The scope
+     * is automatically removed once the operation finishes or throws.
+     *
+     * @param callable $callback The callback to be executed
+     */
+    public function withScope(callable $callback): void
+    {
+        $scope = $this->pushScope();
+
+        try {
+            $callback($scope);
+        } finally {
+            $this->popScope();
+        }
+    }
+
+    /**
+     * Calls the given callback passing to it the current scope so that any
+     * operation can be run within its context.
+     *
+     * @param callable $callback The callback to be executed
+     */
+    public function configureScope(callable $callback): void
+    {
+        $layer = $this->getStackTop();
+
+        if (null !== $layer->getClient()) {
+            $callback($layer->getScope());
+        }
+    }
+
+    /**
+     * Binds the given client to the current scope.
+     *
+     * @param ClientInterface $client The client
+     */
+    public function bindClient(ClientInterface $client): void
+    {
+        $layer = $this->getStackTop();
+        $layer->setClient($client);
+    }
+
+    /**
+     * Captures a message event and sends it to Sentry.
+     *
+     * @param string   $message The message
+     * @param Severity $level   The severity level of the message
+     *
+     * @return null|string
+     */
+    public function captureMessage(string $message, ?Severity $level = null): ?string
+    {
+        $client = $this->getClient();
+
+        if (null !== $client) {
+            return $this->lastEventId = $client->captureMessage($message, [], ['level' => $level]);
+        }
+
+        return null;
+    }
+
+    /**
+     * Captures an exception event and sends it to Sentry.
+     *
+     * @param \Throwable $exception The exception
+     *
+     * @return null|string
+     */
+    public function captureException(\Throwable $exception): ?string
+    {
+        $client = $this->getClient();
+
+        if (null !== $client) {
+            return $this->lastEventId = $client->captureException($exception);
+        }
+
+        return null;
+    }
+
+    /**
+     * Records a new breadcrumb which will be attached to future events. They
+     * will be added to subsequent events to provide more context on user's
+     * actions prior to an error or crash.
+     *
+     * @param Breadcrumb $breadcrumb The breadcrumb to record
+     */
+    public function addBreadcrumb(Breadcrumb $breadcrumb): void
+    {
+        $client = $this->getClient();
+
+        if (null !== $client) {
+            $client->leaveBreadcrumb($breadcrumb);
```

Comment:
This should be done through the scope directly, right?

Reply:
Theoretically yes, but looking at the source code of the Javascript SDK they [call the client passing to it the scope](https://github.com/getsentry/sentry-javascript/blob/65965ff30b4c978c4f200ef8bb5c88d220059324/packages/hub/src/hub.ts#L65), do some operations I don't really understand and then add the breadcrumb to the given scope. I've changed this to directly push the breadcrumb to the scope but I would like an opinion on the matter from @HazAT

Reply:
So we pass it through the Client because the options do live on the Client.
One of the options we support is `beforeBreadcrumb`. `beforeBreadcrumb` receives the breadcrumb before it will be added, user can decide if they want to "add" it or "drop" it.
The Client also receives the current scope on `leaveBreadcrumb` (which should also be renamed to `addBreadcrumb`) and adds it to the `Scope` if `beforeBreadcrumb` says so.